transmission_interface/include/transmission_interface/differential_transmission.hpp

```diff
@@ -0,0 +1,403 @@
+   * \pre Actuator and joint vectors must have size 2 and point to valid data.
+   *  To call this method it is not required that all other data vectors contain valid data, and can even remain empty.
+   */
+  void actuator_to_joint() override;
+
+  /**
+   * \brief Transform variables from joint to actuator space.
+   * \pre Actuator and joint vectors must have size 2 and point to valid data.
+   *  To call this method it is not required that all other data vectors contain valid data, and can even remain empty.
+   */
+  void joint_to_actuator() override;
+
+  std::size_t num_actuators() const override {return 2;}
+  std::size_t num_joints()    const override {return 2;}
+
+  const std::vector<double> & get_actuator_reduction() const {return actuator_reduction_;}
+  const std::vector<double> & get_joint_reduction()    const {return joint_reduction_;}
+  const std::vector<double> & get_joint_offset()       const {return joint_offset_;}
+
+  /**
+   * \brief Get human-friendly report of handles
+   */
+  std::string get_handles_info() const;
+
+protected:
+  std::vector<double> actuator_reduction_;
+  std::vector<double> joint_reduction_;
+  std::vector<double> joint_offset_;
+
+  std::vector<JointHandle> joint_position_;
+  std::vector<JointHandle> joint_velocity_;
+  std::vector<JointHandle> joint_effort_;
+
+  std::vector<ActuatorHandle> actuator_position_;
+  std::vector<ActuatorHandle> actuator_velocity_;
+  std::vector<ActuatorHandle> actuator_effort_;
+};
+
+inline DifferentialTransmission::DifferentialTransmission(
+  const std::vector<double> & actuator_reduction,
+  const std::vector<double> & joint_reduction,
+  const std::vector<double> & joint_offset)
+: actuator_reduction_(actuator_reduction),
+  joint_reduction_(joint_reduction),
+  joint_offset_(joint_offset)
+{
+  if (num_actuators() != actuator_reduction_.size() ||
+    num_joints() != joint_reduction_.size() ||
+    num_joints() != joint_offset_.size())
+  {
+    throw Exception("Reduction and offset vectors must have size 2.");
+  }
+
+  if (0.0 == actuator_reduction_[0] ||
+    0.0 == actuator_reduction_[1] ||
+    0.0 == joint_reduction_[0] ||
+    0.0 == joint_reduction_[1])
+  {
+    throw Exception("Transmission reduction ratios cannot be zero.");
+  }
+}
+
+
+template<typename T>
+std::string to_string(const std::vector<T> & list)
+{
+  std::stringstream ss;
+  ss << "[";
+  for (const auto & elem : list) {
+    ss << elem << ", ";
+  }
+
+  if (!list.empty()) {
+    ss.seekp(-2, std::ios_base::end);  // remove last ", "
+  }
+  ss << "]";
+  return ss.str();
+}
+
+template<class T>
+std::vector<std::string> get_names(const std::vector<T> & handles)
+{
+  std::set<std::string> names;
+  std::transform(
+    handles.cbegin(), handles.cend(), std::inserter(names, names.end()),
+    [](const auto & handle) {return handle.get_name();});
+  return std::vector<std::string>(names.begin(), names.end());
+}
+
+template<typename T>
+std::vector<T> get_ordered_handles(
```

Comment:
Isn't this something you used in other transmission implementations also? If so, why not move it into `transmission.hpp`?

Reply:
I've refactored this into a new file in https://github.com/ros-controls/ros2_control/pull/307